janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/StandardJanusGraphTx.java

```diff
@@ -798,14 +798,13 @@ public JanusGraphVertexProperty addProperty(VertexProperty.Cardinality cardinali
 //                }
 //            }
 
+            long propId = id == null ? IDManager.getTemporaryRelationID(temporaryIds.nextID()) : id;
+            StandardVertexProperty prop = new StandardVertexProperty(propId, key, (InternalVertex) vertex, normalizedValue, ElementLifeCycle.New);
+            if (config.hasAssignIDsImmediately() && id == null) graph.assignID(prop);
+
```

Comment:
Why did you move this logic from bottom to up?

Reply:
because we need to set `prop.setUpsert(true);` later